services/dedup/scylla/scylla.go

```diff
@@ -0,0 +1,139 @@
+package scylla
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/gocql/gocql"
+	"github.com/samber/lo"
+
+	"github.com/rudderlabs/rudder-go-kit/config"
+	"github.com/rudderlabs/rudder-go-kit/stats"
+	"github.com/rudderlabs/rudder-server/services/dedup/types"
+)
+
+type ScyllaDB struct {
+	scylla *gocql.Session
+	conf   *config.Config
+	stat   stats.Stats
+	// Time to live in seconds for an entry in the DB
+	ttl int
+	// Maximum number of keys to commit in a single batch
+	batchSize      int
+	createTableMu  sync.Mutex
+	createTableMap map[string]*sync.Once
+
+	cacheMu  sync.Mutex
+	cache    map[string]types.KeyValue
+	keyspace string
+}
+
+func (d *ScyllaDB) Close() {
+	d.scylla.Close()
+}
+
+func (d *ScyllaDB) Get(kv types.KeyValue) (bool, int64, error) {
+	// Create the table if it doesn't exist
+	var err error
+	d.createTableMu.Lock()
+	once, found := d.createTableMap[kv.WorkspaceId]
+	if !found {
+		d.createTableMap[kv.WorkspaceId] = &sync.Once{}
+		once = d.createTableMap[kv.WorkspaceId]
+	}
+	d.createTableMu.Unlock()
+	once.Do(func() {
+		query := fmt.Sprintf("CREATE TABLE IF NOT EXISTS %s.%s (id text PRIMARY KEY,size bigint) WITH bloom_filter_fp_chance = 0.005;", d.keyspace, kv.WorkspaceId)
+		err = d.scylla.Query(query).Exec()
+	})
+	if err != nil {
```

Comment:
You can use [sync.OnceValue](https://pkg.go.dev/sync#OnceValue) to store and reuse the error.

In the existing code, in case of an error, only the first goroutine will return the error. 
```suggestion
		d.createTableMap[kv.WorkspaceId] = &sync.OnceValue[error]{}
		once = d.createTableMap[kv.WorkspaceId]
	}
	d.createTableMu.Unlock()

	err := once.Do(func() error {
		query := fmt.Sprintf("CREATE TABLE IF NOT EXISTS %s.%s (id text PRIMARY KEY,size bigint) WITH bloom_filter_fp_chance = 0.005;", d.keyspace, kv.WorkspaceId)
		return d.scylla.Query(query).Exec()
	})
	if err != nil {
```

Reply:
Also, you might want to move this logic to a separate unexported method, and call it even from `Commit`. It depends on how robust you want to make your API, normally Get for a workspace should always happen before a commit.